### PR TITLE
BUG: Fix outline merge ordering when inserting at specific position

### PR DESCRIPTION
**Pull request by adityamoolya**
Closes TODO at pypdf/_writer.py line 2792.

The TODO was about using the before parameter when inserting outlines during merge(). Previously, when pages from another PDF were merged at a specific position, their bookmarks were always added to the end of the outline instead of being inserted at the correct position.

For example, merging Doc B into Doc A at position=1 should result in [A1, B1, B2, A2, A3], but previously produced [A1, A2, A3, B1, B2].

The main issue was in TreeObject.insert_child() in _data_structures.py. The traversal started from /Last and followed /Next. Since /Last has no /Next, the traversal stopped immediately and the code always fell back to appending the new outline items.

The fix changes the traversal to start from /First and follow /Next, allowing the correct insertion point to be found. The before parameter is also passed through to _insert_filtered_outline() so the bookmarks are inserted at the same position as the merged pages.

I used Claude Opus 4.6 to write tests for the same

### CODE REVIEW

File: pypdf/generic/_data_structures.py [hunk 1]
```diff
@@ -825,24 +826,27 @@ def insert_child(
...
-            inc_parent_counter(self, child_obj.get("/Count", 1))
-            return child_reference
-        try:  # insert as first or in the middle
-            assert isinstance(prev["/Prev"], DictionaryObject)
-            prev["/Prev"][NameObject("/Next")] = child_reference
-            child_obj[NameObject("/Prev")] = prev["/Prev"]
-        except Exception:  # it means we are inserting in first position
-            child_obj.pop("/Next", None)
-        child_obj[NameObject("/Next")] = prev
+            else:  # `before` not found; append at the end
```
Comment (stefan6419846) [inline pypdf/generic/_data_structures.py]: During my last change to these lines, it has been an explicit choice to use an early return (here more an "early continue") to avoid unnecessary branching/indentation.

**Review by stefan6419846 — CHANGES_REQUESTED** [review]
Thanks for the PR, although I have not fully reviewed it due to its complexity.

At the moment, coverage is quite bad. Could you please review it and make sure that all relevant lines are covered correctly.

File: pypdf/_writer.py [hunk 2]
```diff
@@ -3161,6 +3176,49 @@ def _clone_outline(self, dest: Destination) -> TreeObject:
             n_ol[NameObject("/C")] = ArrayObject(color)
         return n_ol
 
+    def _find_outline_item_before_page(
```
Comment (stefan6419846) [inline pypdf/_writer.py]: I guess moving this to `pypdf.generic._outline` makes more sense to not bloat the large file even more.

File: pypdf/_writer.py [hunk 3]
```diff
@@ -3161,6 +3176,49 @@ def _clone_outline(self, dest: Destination) -> TreeObject:
...
+        self,
+        page_number: int,
+        parent: TreeObject,
+    ) -> Union[TreeObject, IndirectObject, None]:
+        """
+        Find the first top-level outline child of *parent* whose destination
+        page index is >= *page_number*.
+
+        Returns the child's ``IndirectObject`` (suitable for
```
Comment (stefan6419846) [inline pypdf/_writer.py]: This should be a proper docstring.

**Review by stefan6419846 — CHANGES_REQUESTED** [review]
Please check the updated remarks and the merge conflict.

File: pypdf/_writer.py (hunk 2 above)
Comment (adityamoolya) [reply pypdf/_writer.py]: Done , moved to pypdf/generic/_outline.py

File: pypdf/_writer.py (hunk 3 above)
Comment (adityamoolya) [reply pypdf/_writer.py]: Done , converted to a proper docstring .

**Comment by adityamoolya** [comment]
Addressed all review comments , moved _find_outline_item_before_page to pypdf/generic/_outline.py and  added a proper docstring, and resolved the merge conflict. CI is passing as well .

File: pypdf/generic/_outline.py [hunk 4]
```diff
@@ -31,3 +35,51 @@ def write_to_stream(
...
+            if isinstance(action, DictionaryObject) and action.get("/S") == "/GoTo":
+                d = action.get("/D")
+                if d is not None:
+                    d = d.get_object()
+                if isinstance(d, ArrayObject) and len(d) > 0:
+                    page_ref = d[0]
+        if page_ref is None:
+            continue
+        try:
+            pn = writer._get_page_number_by_indirect(page_ref)
```
Comment (stefan6419846) [inline pypdf/generic/_outline.py]: Please avoid such abbreviations.

File: pypdf/_writer.py [hunk 5]
```diff
@@ -683,8 +684,18 @@ def _get_page_number_by_indirect(
         assert indirect_reference is not None, "mypy"
         if isinstance(indirect_reference, int):
             indirect_reference = IndirectObject(indirect_reference, 0, self)
+        target_id = getattr(indirect_reference, "idnum", None)
+        if target_id is not None:
+            for idx, page in enumerate(self.pages):
```
Comment (stefan6419846) [inline pypdf/_writer.py]: This will quickly become a bottleneck for files with lots of outlines. I guess we need some sort of cache object numbers to page indizes.

File: pypdf/_writer.py [hunk 6]
```diff
@@ -683,8 +684,18 @@ def _get_page_number_by_indirect(
...
+                if page.indirect_reference is not None and page.indirect_reference.idnum == target_id:
```
Comment (stefan6419846) [inline pypdf/_writer.py]: Isn't this behavior incorrect if we are dealing with multiple generations?

File: pypdf/_writer.py [hunk 7]
```diff
@@ -683,8 +684,18 @@ def _get_page_number_by_indirect(
...
+                    return idx
         obj = indirect_reference.get_object()
         if isinstance(obj, PageObject):
+            if obj.indirect_reference is not None:
+                target_id = obj.indirect_reference.idnum
+                for idx, page in enumerate(self.pages):
```
Comment (stefan6419846) [inline pypdf/_writer.py]: This is the same as above. Can we avoid the repetition?

File: pypdf/generic/_outline.py [hunk 8]
```diff
@@ -31,3 +35,51 @@ def write_to_stream(
...
+        except Exception as exc:
+            logger_warning(
+                "Could not resolve page number for outline item: %(exc)s",
+                source=__name__,
+                exc=exc,
+            )
+            continue
+        if pn is not None and pn >= page_number:
```
Comment (stefan6419846) [inline pypdf/generic/_outline.py]: Doesn't this assume that outline items are ordered by destination page number, which we cannot assume?

File: pypdf/generic/_outline.py [hunk 9]
```diff
@@ -31,3 +35,51 @@ def write_to_stream(
...
+    Args:
+        page_number: The page index to compare against.
+        parent: The TreeObject representing the parent outline node.
+        writer: The PdfWriter object used to resolve page references.
+
+    Returns:
+        The child's ``IndirectObject`` (suitable for the *before* parameter of
+        ``TreeObject.insert_child``), or ``None`` if no such child exists.
+    """
+    for child in parent.children():
```
Comment (stefan6419846) [inline pypdf/generic/_outline.py:105]: Why can we omit nested outlines?

File: pypdf/generic/_outline.py [hunk 10]
```diff
@@ -31,3 +35,51 @@ def write_to_stream(
...
+                d = action.get("/D")
+                if d is not None:
+                    d = d.get_object()
+                if isinstance(d, ArrayObject) and len(d) > 0:
+                    page_ref = d[0]
+        if page_ref is None:
+            continue
+        try:
+            pn = writer._get_page_number_by_indirect(page_ref)
+        except Exception as exc:
```
Comment (stefan6419846) [inline pypdf/generic/_outline.py]: Which exceptions do we actually expect here?

File: tests/test_merger.py [hunk 11]
```diff
@@ -536,6 +548,252 @@ def test_merge__null_destination():
...
+    buf_merged = BytesIO()
+    merged.write(buf_merged)
+    buf_merged.seek(0)
+    reader = PdfReader(buf_merged)
+
+    titles = [el.title for el in reader.outline if isinstance(el, Destination)]
+    assert titles == ["B1", "A_Dest"]
+
+
+def test_merge_outline_ordering_coverage_edge_cases():
```
Comment (stefan6419846) [inline tests/test_merger.py]: Please use focused tests and consider moving them to `tests/generic/test_outline.py`. The current test setup is rather complex and very implementation-driven.

**Review by stefan6419846 — CHANGES_REQUESTED** [review]

File: pypdf/generic/_outline.py (hunk 4 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py]: Alright, renamed the abbreviations and extracted the destination parsing into a helper to make _find_outline_item_before_page cleaner.

File: pypdf/_writer.py (hunk 5 above)
Comment (adityamoolya) [reply pypdf/_writer.py]: Added a page-ID cache to avoid repeated scans and improve performance

File: pypdf/_writer.py (hunk 6 above)
Comment (adityamoolya) [reply pypdf/_writer.py]: In PdfWriter, object generations are always 0, so idnum should be sufficient for the cache. I also kept the object-resolution fallback if the cache lookup misses.

File: pypdf/generic/_outline.py (hunk 8 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py]: This now scans all direct children and selects the smallest destination page number that is still >= page_number, so it no longer assumes outline order matches page order.

File: pypdf/generic/_outline.py (hunk 9 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py:105]: insert_child operates on the parent's direct /First -> /Next sibling chain, so the before reference must be a direct child of that parent. A nested item's reference would not be found, which is why only direct children are considered

File: pypdf/generic/_outline.py (hunk 10 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py]: I've narrowed the broad exception handling to exactly except (PdfReadError, ValueError):: . These are the specific exceptions that can realistically occur while resolving the page reference (e.g., PdfReadError if the object is missing, or ValueError if the reference belongs to a different PdfWriter context).

File: tests/test_merger.py (hunk 11 above)
Comment (adityamoolya) [reply tests/test_merger.py]: Added focused unit tests in tests/generic/test_outline.py and replaced the old monolithic edge-case test with individual focused tests.

File: pypdf/generic/_outline.py (hunk 8 above)
Comment (stefan6419846) [reply pypdf/generic/_outline.py]: Can you please clarify whether the following LLM-based claim about the semantic outline change holds true and whether this is what we really want?

> Suppose the existing outline order is:
> 
> ```
> Chapter A -> page 20
> Chapter B -> page 5
> Chapter C -> page 30
> ```
> 
> and you're inserting something at page 10.
> 
> Your algorithm selects:
> 
> ```
> Chapter A -> page 20
> ```
> 
> because 20 is the smallest destination >= 10.
> 
> So the resulting order becomes:
> 
> ```
> Chapter B (5)
> New item (10)
> Chapter A (20)
> Chapter C (30)
> ```
> 
> which happens to sort those particular items by page - but it has changed the semantic ordering of the existing outline.
> 
> The whole purpose of before should arguably be:
> 
> > "Where in the existing outline tree should an item associated with this page be inserted?"
> 
> not:
> 
> > "Find the outline item whose page is numerically closest above this page."

File: pypdf/generic/_outline.py (hunk 9 above)
Comment (stefan6419846) [reply pypdf/generic/_outline.py:105]: Let's say we have:

* Chapter 1 pointing to page 1
* Section 1.1 pointing to page 20
* Chapter 2 pointing to page 30

When merging with page 10 now, we would currently select "merge before chapter 2", although from a structural point of view one could argue that it somehow belongs between section 1.1 and chapter 1.

File: pypdf/generic/_outline.py [hunk 12]
```diff
@@ -31,3 +36,94 @@ def write_to_stream(
...
+
+    Returns:
+        The page indirect reference, or ``None`` if it cannot be resolved.
+
+    """
+    if "/Dest" in child:
+        dest = child["/Dest"].get_object()
+        if isinstance(dest, ArrayObject) and len(dest) > 0:
+            return cast(IndirectObject, dest[0])
+    elif "/A" in child:
```
Comment (stefan6419846) [inline pypdf/generic/_outline.py:61]: This assumes that only one value is possible, which the specification states, but if the destination is a `NullObject` for example, I would argue that we can try the `/A` key as well.

File: pypdf/_writer.py [hunk 13]
```diff
@@ -683,8 +710,19 @@ def _get_page_number_by_indirect(
...
+        if _page_cache is None:
+            _page_cache = self._build_page_id_cache()
+        target_id = getattr(indirect_reference, "idnum", None)
+        if target_id is not None:
+            result = _page_cache.get(target_id)
+            if result is not None:
+                return result
         obj = indirect_reference.get_object()
         if isinstance(obj, PageObject):
+            if obj.indirect_reference is not None:
```
Comment (stefan6419846) [inline pypdf/_writer.py:722]: When inverting the condition, we can move the first nested block one level up.

File: pypdf/generic/_outline.py [hunk 14]
```diff
@@ -31,3 +36,94 @@ def write_to_stream(
...
+        page_ref = _resolve_outline_dest_page_ref(child)
+        if page_ref is None:
+            continue
+        try:
+            child_page_number = writer._get_page_number_by_indirect(
+                page_ref, _page_cache=page_cache
+            )
+        except (PdfReadError, ValueError):
+            logger_warning(
+                "Could not resolve page number for outline item",
```
Comment (stefan6419846) [inline pypdf/generic/_outline.py:115]: I would at least include the reason here in case we somehow miss a legitimate case.

File: tests/generic/test_data_structures.py [hunk 15]
```diff
@@ -541,6 +546,81 @@ def test_tree_object__insert_child__cycle() -> None:
...
+    existing = TreeObject()
+    existing_ref = writer._add_object(existing)
+    tree3.insert_child(existing_ref, None, writer)
+
+    child3 = TreeObject()
+    child3[NameObject("/Prev")] = NullObject()
+    child3_ref = writer._add_object(child3)
+    tree3.insert_child(child3_ref, existing_ref, writer)
+    assert "/Prev" not in child3
+
```
Comment (stefan6419846) [inline tests/generic/test_data_structures.py:623]: Functions should be divided by exactly two empty lines.

File: tests/generic/test_outline.py [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+    NameObject,
+    TextStringObject,
+    TreeObject,
+)
+from pypdf.generic._outline import (
+    _find_outline_item_before_page,
+    _resolve_outline_dest_page_ref,
+)
+
+# ── helpers ──────────────────────────────────────────────────────────
```
Comment (stefan6419846) [inline tests/generic/test_outline.py:21]: These comments are not required.

File: tests/generic/test_outline.py [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+        NameObject("/S"): NameObject("/GoTo"),
+        NameObject("/D"): ArrayObject([page_ref, NameObject("/Fit")]),
+    })
+    child = DictionaryObject()
+    child[NameObject("/A")] = action
+
+    assert _resolve_outline_dest_page_ref(child) == page_ref
+
+
+def test_resolve_returns_none_for_no_dest() -> None:
```
Comment (stefan6419846) [inline tests/generic/test_outline.py:71]: I would avoid keeping the expected outcome in the test name. The test name should IMHO cover what is tested, for example that the `/Dest` entry is missing.


```suggestion
def test_resolve_outline_destination_page_reference__no_dest_entry() -> None:
```

**Review by stefan6419846 — CHANGES_REQUESTED** [review]
While the tests attempt to cover all edge cases from the code, some real-world use cases might be missing:

* Existing outline pages are out of order
* Duplicate destinations
* Nested outlines
* Item with `/Dest` and `/A`
* Named destinations
* Merging into an empty outline
* Merging before the first outline
* Merging after the last outline
* Merging where no bookmark exists after the insertion page
* Inserting exactly at a bookmark's page
* Multiple existing bookmarks on the same page

File: pypdf/generic/_outline.py (hunk 12 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py]: Changed the elif "/A" in child: check to a non-exclusive if "/A" in child: block. If /Dest exists but does not resolve to a valid destination array (such as a NullObject) , it now falls through to inspect /A as well. 
Also added a dedicated test case test_resolve_dest_null_falls_through_to_action for this scenario.

File: pypdf/generic/_outline.py (hunk 14 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py]: Now it captures the exception via except (PdfReadError, ValueError) as exc: and  includes the error details in the message

File: pypdf/_writer.py (hunk 13 above)
Comment (adityamoolya) [reply pypdf/_writer.py]: Inverted the condition to an early return:

```

obj = indirect_reference.get_object()
if not isinstance(obj, PageObject):
    return None
if obj.indirect_reference is not None:
    result = _page_cache.get(obj.indirect_reference.idnum)
    if result is not None:
        return result
return obj.page_number 
```
This eliminates the extra level of indentation.

File: tests/generic/test_data_structures.py (hunk 15 above)
Comment (adityamoolya) [reply tests/generic/test_data_structures.py:622]: resolved

File: tests/generic/test_outline.py (hunk 16 above)
Comment (adityamoolya) [reply tests/generic/test_outline.py]: removed them

**Comment by adityamoolya** [comment]
> While the tests attempt to cover all edge cases from the code, some real-world use cases might be missing:
> 
> * Existing outline pages are out of order
> * Duplicate destinations
> * Nested outlines
> * Item with `/Dest` and `/A`
> * Named destinations
> * Merging into an empty outline
> * Merging before the first outline
> * Merging after the last outline
> * Merging where no bookmark exists after the insertion page
> * Inserting exactly at a bookmark's page
> * Multiple existing bookmarks on the same page

Added  test coverage in `tests/generic/test_outline.py` and `tests/generic/test_data_structures.py` covering all requested scenarios:
- Out-of-order outline page numbers (`test_merge_outline_ordering_out_of_order`)
- Duplicate destinations on identical pages (`test_merge_outline_duplicate_destinations`)
- Multiple bookmarks on the same page (`test_merge_multiple_bookmarks_on_same_page`)
- Merging into empty outlines (`test_merge_into_empty_outline`)
- Merging when no bookmark exists after the target position (`test_merge_where_no_bookmark_after_insertion_page`)
- Nested outline trees (`test_merge_outline_with_nested_outlines`)
- Middle insertions in linked-list tree objects (`test_tree_object__insert_child_cleanup_coverage` case 4)

File: pypdf/generic/_outline.py (hunk 8 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py]: yes, I think the LLM's analysis is correct for a non-monotonic outline. with the current implementation, we look at all direct children and choose the one with the smallest destination page >= `page_number`, so the insertion is based on page order rather than the existing sibling order.

my original intention was to fix the case where the merged outline was always appended at the end, without assuming that the existing outline is already sorted by page number.

but I agree this raises the question of what ordering we actually want here. if we want to preserve the existing sibling order, we could instead stop at the first child encountered with a destination page >= `page_number`. that would avoid moving the insertion point based on a later sibling.

i'm not sure which behavior is the intended one for a non-monotonic outline, though. what do you think `before` should mean in this case?

File: pypdf/generic/_outline.py (hunk 9 above)
Comment (adityamoolya) [reply pypdf/generic/_outline.py:97]: the reason i only consider direct children is that `TreeObject.insert_child()` operates on the parent's direct `/First -> /Next` sibling chain, so a nested item wouldn't be a valid `before` reference for that parent.

i also wasn't sure it would be correct to automatically infer that the imported outline should become part of an existing nested section. for example, in the Chapter 1 / Section 1.1 case, deciding that page 10 belongs inside Chapter 1 would require us to infer the intended scope of the existing outline from the page destinations.

if the user explicitly wants the imported outline under a particular existing outline item, `outline_item` already provides that parent.

so my current thinking was to keep the automatic position calculation at the direct-child level.

File: tests/generic/test_outline.py (hunk 17 above)
Comment (adityamoolya) [reply tests/generic/test_outline.py]: applied suggestion
